src/servers/grpc_server.cc

```diff
@@ -2806,20 +2858,34 @@ ModelInferHandler::InferResponseComplete(
     TRITONSERVER_InferenceResponse* iresponse, const uint32_t flags,
     void* userp)
 {
-  // FIXME need to correctly handle FINAL flag
-  if (iresponse == nullptr) {
-    return;
-  }
-
   State* state = reinterpret_cast<State*>(userp);
 
+  // Increment the callback index
+  state->cb_counter_++;
+
   LOG_VERBOSE(1) << "ModelInferHandler::InferResponseComplete, "
                  << state->unique_id_ << " step " << state->step_;
 
+  // Defer to the callback with the final response
+  if ((flags & TRITONSERVER_RESPONSE_COMPLETE_FINAL) == 0) {
```

Comment:
For ModelInfer we expect there to be a single response and it must have the FINAL flag. So if the final flag is not given we should log an error. Or maybe the below code handle it?

Reply:
The below code will correctly handle the cases where model produced multiple responses for ModelInfer if ModelConfig didn't specify decoupled policy. It just keeps on deferring the return of actual response. Nevertheless, it;'s better to log an error because may be the single response doesn't carry the specify the FINAL flag.